anonlink/solving/_multiparty_solving_inner.cpp

```diff
@@ -0,0 +1,254 @@
+#include <cassert>
+#include <unordered_map>
+#include <unordered_set>
+#include "_multiparty_solving_inner.h"
+
+struct RecordHasher {
+    std::hash<int> hasher;
+    size_t operator()(const Record &record) const {
+        size_t hash1 = hasher(record.dset_i);
+        size_t hash2 = hasher(record.rec_i);
+
+        // https://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html
+        return hash1 ^ (hash2 + 0x9e3779b9 + (hash1 << 6) + (hash1 >> 2));
+    }
+};
+
+
+
+class GroupsStore {
+// Keep track of groups. This is like a low-effort Disjoint Sets data structure.
+private:
+    typedef std::unordered_map<Record, Group*, RecordHasher> InnerStore;
+    InnerStore inner_store;
+    
+#ifndef NDEBUG
+    bool in_group(Record record) const {
+        return this->inner_store.count(record);
+    }
+#endif /* NDEBUG */
+    
+public:
+    Group *get_group(Record record) const {
+        auto res_iterator = this->inner_store.find(record);
+        if (res_iterator == this->inner_store.end()) {
+            return nullptr;
+        } else {
+            Group *retval = res_iterator->second;
+            assert(retval);
+            return retval;
+        }
+    }
+    
+    Group *make_group(Record record) {
+        assert(!this->in_group(record));
+        Group *group = new Group {record};
+        this->inner_store[record] = group;
+        return group;
+    }
+    
+    Group *make_group(Record record0, Record record1) {
+        assert(!this->in_group(record0));
+        assert(!this->in_group(record1));
+        Group *group = new Group {record0, record1};
+        this->inner_store[record0] = group;
+        this->inner_store[record1] = group;
+        return group;
+    }
+    
+    Group *add_to_group(Group *group, Record record) {
+        assert(!this->in_group(record));
+        assert(group);
+        assert(std::all_of(group->cbegin(),
+                           group->cend(),
+                           [this, group](Record r) { return this->inner_store.at(r) == group; }));
+        group->push_back(record);
+        this->inner_store[record] = group;
+        return group;
+    }
+    
+    Group *merge_into(Group * __restrict__ absorber, Group * __restrict__ absorbee) {
+        assert(absorber);
+        assert(absorbee);
+        assert(std::all_of(absorber->cbegin(),
```

Comment:
`std::all_of` is in the `<algorithm>` header which you need to `#include` at the top.